File: boot_images/core/buses/uart/src/Uart_log.c
#if ( defined(UART_LOG_ALL) || defined(UART_LOG_ERROR) )
/*==================================================================================================

FILE: Uart_log.c

DESCRIPTION: This module provides the logging feature for the UART driver Software.

                           Copyright (c) 2013 Qualcomm Technologies Incorporated
                                        All Rights Reserved
                                     QUALCOMM Proprietary/GTDR

==================================================================================================*/

/*==================================================================================================
Edit History

$Header: //components/rel/boot.bf/3.1.2.c3/boot_images/core/buses/uart/src/Uart_log.c#1 $

when       who     what, where, why
--------   ---     --------------------------------------------------------
07/10/13   VV      Initial revision.

==================================================================================================*/

/*==================================================================================================
                                           INCLUDE FILES
==================================================================================================*/
#include "Uart_log.h"


/*==================================================================================================
                                          LOCAL VARIABLES
==================================================================================================*/
#define LOG_BUF_SIZE 512
#define MSG_HEADER_SIZE (sizeof(MSG_HEADER))
#define MSG_START 0x1234ABCD

typedef struct
{
   uint32  msg_start;
   uint16  file_id;
   uint16  line;
   uint16  num_args;
   uint16  msg_seq;
}MSG_HEADER;

typedef struct
{
   uint8* ptr;
   uint8  log[LOG_BUF_SIZE];
   uint16 bytes_left;
   uint16 next_msg_seq;
}UART_LOG_STRUCT;

UART_LOG_STRUCT uart_log_struct =
{
   NULL,
   {0},
   LOG_BUF_SIZE,
   0
};


/*==================================================================================================
                                     LOCAL FUNCTION PROTOTYPES
==================================================================================================*/
static void fill_msg_header(MSG_HEADER* msg_ptr, uint8 file_id, uint16 line, uint8 num_args);

/*==================================================================================================
                                          LOCAL FUNCTIONS
==================================================================================================*/
static void fill_msg_header(MSG_HEADER* msg_ptr,uint8 file_id,uint16 line, uint8 num_args)
{

   msg_ptr->msg_start = MSG_START;
   msg_ptr->file_id   = file_id;
   msg_ptr->line      = line;
   msg_ptr->num_args  = num_args;
   msg_ptr->msg_seq   = uart_log_struct.next_msg_seq++ ;

}

/*==================================================================================================
                                          GLOBAL FUNCTIONS
==================================================================================================*/
void uart_log(uint8 file_id,uint16 line, uint8 num_args, uint32 a1, uint32 a2, uint32 a3)
{
   uint32   msg_bytes = 0;
   uint32   args[3];
   uint32*  data_ptr;
   uint8    i;
   uint8*   ptr;
   if(num_args >= 3)
   {
      return;
   }

   args[0] = a1;
   args[1] = a2;
   args[2] = a3;

   // Calculate the total msg size.
   msg_bytes = MSG_HEADER_SIZE + num_args * sizeof( uint32 );

   // If it is the first time the UART_LOG is called or the buffer is full, roll over.
   if ((uart_log_struct.ptr == NULL) || (msg_bytes > uart_log_struct.bytes_left))
   {
      uart_log_struct.ptr = uart_log_struct.log;
      uart_log_struct.bytes_left = LOG_BUF_SIZE;
   }

   uart_log_struct.bytes_left -= msg_bytes;
   ptr = uart_log_struct.ptr;
   uart_log_struct.ptr += msg_bytes;

   // Fill in the header
   fill_msg_header((MSG_HEADER*)ptr,file_id,line,num_args);
   ptr = ptr + MSG_HEADER_SIZE;
   data_ptr = (uint32*) ptr;

   // Fill in the data
   for (i = 0;i < num_args; i++)
   {
      *data_ptr++ = args[i];
   }
}

#endif

File: modem_proc/core/buses/uart_lite/src/uart_log.py
"""
/*===================================================================================================

FILE: Uart_log.py

DESCRIPTION: This module provides the script to decode the UART log file.

USAGE:
    python Uart_log.py <log file> <decoded_txt_file>

                           Copyright (c) 2014 QUALCOMM Incorporated
                                        All Rights Reserved
                                     QUALCOMM Proprietary/GTDR

==================================================================================================*/
/*==================================================================================================
Edit History

$Header: //components/rel/core.mpss/3.5.c12.3/buses/uart_lite/src/uart_log.py#1 $

when       who     what, where, why
--------   ---     --------------------------------------------------------
07/15/13   VV      Initial revision.

==================================================================================================*/
"""

import re
import os
import sys
import fnmatch



def get_byte(log_buf):
    byte = int(log_buf.pop(0),16)
    return byte

def get_word(log_buf):
    word = int(log_buf.pop(0),16)
    word = word | int(log_buf.pop(0),16) << 8
    return word

def get_dword(log_buf):
    dword = int(log_buf.pop(0),16)
    dword = dword | int(log_buf.pop(0),16) << 8
    dword = dword | int(log_buf.pop(0),16) << 16
    dword = dword | int(log_buf.pop(0),16) << 24
    return dword

def get_raw_msg(log_buf):
    msg = {}
    msg_start = ["0xcd","0xab","0x34","0x12"]
    i = 0
    while (i < 4) and (log_buf != []):
        if msg_start[i] != hex(get_byte(log_buf)):
            i = 0
            continue
        else:
            i = i+1

    if i!=4:
        return msg
    msg['file_id']  = get_word(log_buf)
    msg['line']     = get_word(log_buf)
    msg['num_args'] = get_word(log_buf)
    msg['msg_seq']  = get_word(log_buf)
    msg['log_level']= "INVALID"
    return msg

def get_line_content(msg,files):

    file_content = files[msg['file_id']][1]
    file_content = file_content.split("\n")
    args = msg['args']

    try:
        source_line = file_content[msg['line'] - 1]
    except:
        print "Line number error. Line number decoded: " + str(msg['line'] - 1)
        print "File ID : " + str(msg['file_id'])
       

    try:
        msg_string = re.search("(?<=\")(.*)(?=\")",source_line).group(0)
    except:
        msg['log_level'] = "INVALID"
        msg_string = "---------------"
        return msg_string

    try:
        msg['log_level'] = re.search("INFO|ERROR",source_line).group(0)
    except:
        msg['log_level'] = "INVALID"
        msg_string = "---------------"
        return msg_string

    for i in range(0,msg['num_args']):
        msg_string = re.sub("%d",str(args[i]),msg_string,1)
    return  msg_string

def parse_contents(log_buf,files,msgs):

    while log_buf != []:
        raw_msg = get_raw_msg(log_buf)
        if raw_msg == {}:
            return

        args = []
        decoded_msg = {}

        for i in range(0,raw_msg['num_args']):
            args.append(get_dword(log_buf))
        raw_msg['args'] = args

        file_name = re.search("(\w)*\.c",files[raw_msg['file_id']][0],re.I).group(0)

        decoded_msg['msg']          = get_line_content(raw_msg,files)
        decoded_msg['seq']          = raw_msg['msg_seq']
        decoded_msg['file_name']    = file_name
        decoded_msg['line_number']  = raw_msg['line']
        decoded_msg['log_level']    = raw_msg['log_level']
        msgs.append(decoded_msg)

def parse_files(files):
    curr_dir = os.getcwd()
    par_dir = re.search(r"(.*)uart\\",curr_dir).group(0)

    src_files = []
    for root, dirnames, filenames in os.walk(par_dir):
      for filename in fnmatch.filter(filenames, '*.c'):
          src_files.append(os.path.join(root, filename))
    for src in src_files:
        file_content = open(src,"r").read()
        try:
            file_trace_id = int(re.search("UART_LOGGING_FILE_ID (\d*)",file_content).group(1))
            files[file_trace_id] = [src,file_content]
        except:
            file_trace_id = 0
    return files

msgs = []
files = {}
log_file_name = sys.argv[1]
log_file = open(log_file_name,"r")

file_content = log_file.read()
log_file.close()
log_buf = re.search(r"(?<=log = \()(.*?)(?=\))",file_content,re.S)


log_buf = str(log_buf.group(0))
log_buf = log_buf.replace(" ","")
log_buf = log_buf.replace("\n","")
log_buf = log_buf.split(",")

parse_files(files)
parse_contents(log_buf,files,msgs)
sorted(msgs,key = lambda k: k['seq'] )

out_file = open(sys.argv[2],"w")
out_file.write("{:<5} {:^15} {:<15} {:<5} {:<30} \n" .format("Msg Seq.No"," Log Level", "File ", "Line", "Message "))
for i in msgs:
   out_file.write("{:<5} {:^15} {:<15} {:<5} {:<30} \n" .format(str(i['seq']) , i['log_level'] , i['file_name'] , str(i['line_number']) , i['msg']))

out_file.close()

